python/pylibcudf/utils/check-type-stubs.py

```diff
@@ -0,0 +1,484 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+import enum
+import inspect
+import itertools
+import sys
+import types
+import warnings
+from collections.abc import Callable, Generator, Mapping
+from pathlib import Path
+
+import libcst as cst
+import libcst.matchers as m
+
+
+class StubParser(cst.CSTVisitor):
+    def __init__(self):
+        self.classes: list[cst.ClassDef] = []
+        self.functions: list[cst.FunctionDef] = []
+        self.misc: list[cst.AnnAssign] = []
+
+    @classmethod
+    def definitions(
+        cls, file: Path
+    ) -> tuple[list[cst.ClassDef], list[cst.FunctionDef], list[cst.AnnAssign]]:
+        """Extract definitions from a type stub file.
+
+        Parameters
+        ----------
+        file
+            File to extract definitions from
+
+        Returns
+        -------
+        list[ClassDef], list[FunctionDef], list[AnnAssign]
+            Tuple of concrete syntax tree nodes, class definitions,
+            function definitions, annotated assignments.
+        """
+        if file.name == "__init__.py":
+            raise RuntimeError("Not expecting __init__.py here")
+        mod = cst.parse_module(file.read_text())
+        collector = cls()
+        mod.visit(collector)
+        return collector.classes, collector.functions, collector.misc
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> bool:
+        self.classes.append(node)
+        return False
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
+        self.functions.append(node)
+        return False
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool:
+        self.misc.append(node)
+        return False
+
+
+def traverse_modules(
+    root: types.ModuleType,
+) -> Generator[types.ModuleType, None, None]:
+    """
+    Yield unique modules referenced from a root
+
+    Parameters
+    ----------
+    root
+        Root module to start at
+
+    Yields
+    ------
+    Modules referred to by the root module.
+
+    Notes
+    -----
+    This only yields modules named in `__all__`.
+    """
+    lifo = [root]
+    seen = set()
+    while lifo:
+        mod = lifo.pop()
+        members = dict(inspect.getmembers(mod))
+        if "__all__" not in members:
+            raise RuntimeError(f"Module {mod.__name__} must provide __all__")
+        yield mod
+        seen.add(mod)
+        for member in members["__all__"]:
+            obj = members[member]
+            if isinstance(obj, types.ModuleType) and obj not in seen:
+                lifo.append(obj)
+
+
+def check_module(
```

Comment:
Should we should split this function up a little bit? It looks like we can use 3 functions to do:
1. Find `pyi` file corresponding to given module
2. Collect all the functions classes etc.
3. Do the comparisons for funcitions and classes

Reply:
Good idea, thanks.